File: pertemuan2/RekeningBank.java
/**
 * Sesi 3 — constructor berdelegasi, anggota statis, dan konstanta.
 *
 * Invariant:
 *   1. saldo tidak pernah negatif
 *   2. nomor rekening tidak berubah setelah objek dibuat
 *   3. setoran dan penarikan selalu bernilai positif
 */
public class RekeningBank {

    // TODO 1: konstanta
    public static final double BUNGA_TAHUNAN = 0.025;
    public static final double BIAYA_ADMINISTRASI = 5000;
    public static final double BATAS_PENARIKAN_SEKALI = 5000000;

    // TODO 2: field statis penghitung jumlah rekening
    private static int jumlahRekening = 0;

    private final String nomor;
    private final String pemilik;
    private double saldo;

    /**
     * Constructor ringkas.
     * Mendelegasikan ke constructor lengkap.
     */
    public RekeningBank(String nomor, String pemilik) {
        this(nomor, pemilik, 0);
    }

    /** Constructor lengkap — SATU-SATUNYA tempat validasi berada. */
    public RekeningBank(String nomor, String pemilik, double saldoAwal) {

        // TODO 4: validasi nomor dan saldo awal
        if (nomor == null || nomor.isEmpty()) {
            throw new IllegalArgumentException("Nomor rekening tidak boleh kosong");
        }

        if (saldoAwal < 0) {
            throw new IllegalArgumentException("Saldo awal tidak boleh negatif");
        }

        this.nomor = nomor;
        this.pemilik = pemilik;
        this.saldo = saldoAwal;

        // TODO 5: naikkan penghitung jumlah rekening
        jumlahRekening++;
    }

    public void setor(double jumlah) {

        // TODO 6: validasi setoran
        if (jumlah <= 0) {
            throw new IllegalArgumentException("Jumlah setoran harus positif");
        }

        saldo += jumlah;
    }

    public void tarik(double jumlah) {

        // TODO 7: validasi penarikan
        if (jumlah <= 0) {
            throw new IllegalArgumentException("Jumlah penarikan harus positif");
        }

        if (jumlah > saldo) {
            throw new IllegalArgumentException("Saldo tidak mencukupi");
        }

        if (jumlah > BATAS_PENARIKAN_SEKALI) {
            throw new IllegalArgumentException("Jumlah penarikan melebihi batas");
        }

        saldo -= jumlah;
    }

    /** TODO 8: kurangi saldo sebesar biaya administrasi, tetapi jangan sampai negatif. */
    public void potongBiayaAdmin() {
        if (saldo >= BIAYA_ADMINISTRASI) {
            saldo -= BIAYA_ADMINISTRASI;
        } else {
            saldo = 0;
        }
    }

    /** TODO 9: method statis — kembalikan jumlah rekening yang pernah dibuat. */
    public static int getJumlahRekening() {
        return jumlahRekening;
    }

    /**
     * TODO 10: method statis utilitas — hitung bunga setahun dari pokok.
     * Method ini tidak membaca keadaan objek mana pun.
     */
    public static double bungaSetahun(double pokok) {
        return pokok * BUNGA_TAHUNAN;
    }

    public double getSaldo() {
        return saldo;
    }

    public String getNomor() {
        return nomor;
    }

    @Override
    public String toString() {
        return String.format("Rekening[%s] %-14s Rp%,.2f", nomor, pemilik, saldo);
    }
}

File: pertemuan2/RekeningBank.php
<?php
declare(strict_types=1);

/**
 * Sesi 3 — PHP tidak punya constructor overloading.
 * Padanannya: default parameter + named constructor (static factory).
 */
class RekeningBank
{
    // TODO 1: konstanta
    public const BUNGA_TAHUNAN = 0.025;
    public const BIAYA_ADMINISTRASI = 5000;
    public const BATAS_PENARIKAN_SEKALI = 5000000;

    // TODO 2: properti statis penghitung jumlah rekening
    private static int $jumlahRekening = 0;

    private float $saldo;

    /**
     * Default parameter menggantikan constructor overloading.
     * TODO 3: validasi nomor kosong dan saldo awal negatif.
     * TODO 4: naikkan penghitung jumlah rekening.
     */
    public function __construct(
        private readonly string $nomor,
        private readonly string $pemilik,
        float $saldoAwal = 0,
    ) {
        // TODO 3: validasi
        if ($this->nomor === '') {
            throw new InvalidArgumentException(
                'Nomor rekening tidak boleh kosong'
            );
        }

        if ($saldoAwal < 0) {
            throw new InvalidArgumentException(
                'Saldo awal tidak boleh negatif'
            );
        }

        $this->saldo = $saldoAwal;

        // TODO 4: tambah jumlah rekening
        self::$jumlahRekening++;
    }

    /**
     * TODO 5: named constructor — rekening pelajar, saldo awal nol.
     * Gunakan new static(), BUKAN new self().
     */
    public static function rekeningPelajar(
        string $nomor,
        string $pemilik
    ): static {
        return new static($nomor, $pemilik, 0);
    }

    public function setor(float $jumlah): void
    {
        // TODO 6
        if ($jumlah <= 0) {
            throw new InvalidArgumentException(
                'Jumlah setoran harus positif'
            );
        }

        $this->saldo += $jumlah;
    }

    public function tarik(float $jumlah): void
    {
        // TODO 7
        if ($jumlah <= 0) {
            throw new InvalidArgumentException(
                'Jumlah penarikan harus positif'
            );
        }

        if ($jumlah > self::BATAS_PENARIKAN_SEKALI) {
            throw new InvalidArgumentException(
                'Jumlah penarikan melebihi batas'
            );
        }

        if ($jumlah > $this->saldo) {
            throw new InvalidArgumentException(
                'Saldo tidak mencukupi'
            );
        }

        $this->saldo -= $jumlah;
    }

    /** TODO 8 */
    public function potongBiayaAdmin(): void
    {
        if ($this->saldo >= self::BIAYA_ADMINISTRASI) {
            $this->saldo -= self::BIAYA_ADMINISTRASI;
        } else {
            $this->saldo = 0;
        }
    }

    /** TODO 9 */
    public static function getJumlahRekening(): int
    {
        return self::$jumlahRekening;
    }

    /** TODO 10 */
    public static function bungaSetahun(float $pokok): float
    {
        return $pokok * self::BUNGA_TAHUNAN;
    }

    public function getSaldo(): float
    {
        return $this->saldo;
    }

    public function getNomor(): string
    {
        return $this->nomor;
    }

    public function __toString(): string
    {
        return sprintf(
            'Rekening[%s] %-14s Rp%s',
            $this->nomor,
            $this->pemilik,
            number_format($this->saldo, 2, ',', '.')
        );
    }
}